src/core/mac/mac_frame.hpp

```diff
@@ -1144,6 +1176,29 @@ class Frame : public otRadioFrame
     static constexpr uint8_t kMaxPsduSize   = kInvalidSize - 1;
     static constexpr uint8_t kSequenceIndex = kFcfSize;
 
+#if OPENTHREAD_CONFIG_WAKE_UP_COORDINATOR_ENABLE || OPENTHREAD_CONFIG_WAKE_UP_END_DEVICE_ENABLE
+    inline uint16_t FcfDstAddrMask(void) const { return FcfDstAddrMask(GetFrameControlField()); }
+    inline uint16_t FcfSrcAddrMask(void) const { return FcfSrcAddrMask(GetFrameControlField()); }
+    inline uint16_t FcfSecurityEnabledMask(void) const
+    {
+        return IsMultipurpose() ? kFcfMpSecurityEnabled : kFcfSecurityEnabled;
+    }
+
+    inline uint16_t FcfFramePendingMask(void) const { return IsMultipurpose() ? kFcfMpFramePending : kFcfFramePending; }
+    inline uint16_t FcfAckRequestMask(void) const { return IsMultipurpose() ? kFcfMpAckRequest : kFcfAckRequest; }
+    inline uint16_t FcfIePresentMask(void) const { return IsMultipurpose() ? kFcfMpIePresent : kFcfIePresent; }
+#else
+    inline uint16_t        FcfDstAddrMask(void) const { return kFcfDstAddrMask; }
+    inline uint16_t        FcfSrcAddrMask(void) const { return kFcfSrcAddrMask; }
+    inline uint16_t        FcfSecurityEnabledMask(void) const { return kFcfSecurityEnabled; }
+    inline uint16_t        FcfFramePendingMask(void) const { return kFcfFramePending; }
+    inline uint16_t        FcfAckRequestMask(void) const { return kFcfAckRequest; }
+    inline uint16_t        FcfIePresentMask(void) const { return kFcfIePresent; }
+#endif
+
+    Error GetSrcAddrSize(uint8_t &aSize) const { return GetSrcAddrSize(GetFrameControlField(), aSize); };
+    Error GetDstAddrSize(uint8_t &aSize) const { return GetDstAddrSize(GetFrameControlField(), aSize); }
```

Comment:
Thanks @zhanglongxia .

Seeing this PR and the complexity it adds for multi-purpose frame support reminded me of an earlier idea.

Instead of having many methods in `Mac::Frame` for parsing/getting header fields, we could have one method to parse MAC and security headers into a well-structured data object. `Mac` and other layers would then simply use this parsed data. For sending frames, we already have `InitMacHeader()` which does something similar.

I believe this model would simplify multi-purpose frame parsing as well.

I had a branch with a good portion of this implemented (from years ago). I'll try to see if I can find and restore it.

One challenge with this model is the utility lib `utils/mac_frame.h`, which provides `otMacFrame{}` APIs similar to `Mac::Frame` methods (essentially calling them) to aid radio platforms. We need to consider how to accommodate this. 

Thoughts?

Reply:
Thanks @zhanglongxia. Glad you agree that this model would be better! 😃 

I found my old branch, but it needs a fair bit of work. I started on this over the weekend, but it's a larger change and will take more time (there are a lot of places where we use the `Mac::Frame` methods). 

The question is how to proceed. 
* Do we want to move forward with this PR, adding support for multi-purpose frames, and then incorporate it into the new model later? 
* Or should we try to push the new model first and then add multi-purpose frame support on top of it?

Reply:
@abtink How long will it take you to finish modifying this commit in OT? If you can complete the changes within 2 or 3 weeks, we can add multipurpose frame support after you have completed the changes.

Reply:
Great. Thanks. I think less than that for sure. My plan/hope is to have it by end of this week :)